### DISCUSSION TITLE
Laravel Planetscale can migrate but can't query it show SQLSTATE[HY000] [2002]

### DISCUSSION THREAD

**Discussion by SeakMengs** [planetscale/discussion#477] [Bug Reports] [answered]
I'm having a bit of trouble with my Laravel version 8.0 and PlanetScale connection. I've got everything configured in my .env file, including the certificate installation, and I can migrate tables to the database just fine. However, when I try to query the database, I keep getting an error message that says ```Illuminate\Database\QueryException.SQLSTATE[HY000] [2002] (SQL: select * from company_user)```
![image](https://github.com/planetscale/discussion/assets/54373229/a199464d-e56b-4897-b940-19f881bb0987)

These are the .env configuration I copied from planetscale
```
DB_CONNECTION=mysql
DB_HOST=aws.connect.psdb.cloud
DB_PORT=3306
DB_DATABASE=business_directory
DB_USERNAME=<username>
DB_PASSWORD=<password>
//cacert.pem is inside the root directory
MYSQL_ATTR_SSL_CA=cacert.pem
```

I've tried troubleshooting this issue on my own, but so far nothing has worked. I've double-checked all of my configuration settings and made sure that everything is correct, but I still can't seem to get it to work.

Has anyone else experienced this issue before? If so, do you have any suggestions for how to fix it? Any help would be greatly appreciated!

Thanks in advance.

PS. I have tried migrating tables, insert and query data from my localhost it work just fine. It seems that I can test query using "php artisan tinker" and then "DB::table('company_user')->get();" it does return data on the cli but it doesn't work on the controller I'm not sure why.

**Comment by orware** [comment]
Hello @SeakMengs,

I haven't assisted any users that have encountered this particular issue.

I don't think the Laravel version should be a factor here as I don't think much would have changed regarding database connections between 8.x and the current version that would explain that behavior.

The only thing that immediately comes to mind is whether or not the Windows Firewall (or another firewall) may potentially be a factor in disallowing the process from creating the remote connection to the PlanetScale database perhaps?

The connection information seems to all be correct and working since you can run the commands via Tinker, but they only seem to not work when executed via the browser. 

I'm guessing the `http://127.0.0.1:8000/user` URL is being powered by a development web server you are starting on the fly? Investigating the potential blocking due to a firewall would probably be the first suggestion I would look into to start off with.

**Comment by SeakMengs** [comment, marked as answer]
Thank you @orware  for your response regarding the issue with not being able to connect planetscale  on the web. After checking the firewall I found no problem so I decide to set up a virtual machine to test if it work on Linux (Ubuntu to be specific). After minutes of testing I decide to change the SSL certificate path from 
```
MYSQL_ATTR_SSL_CA=cacert.pem 
To
MYSQL_ATTR_SSL_CA=/etc/ssl/certs/ca-certificates.crt.
```

Through this process, I was able to identify that the problem was indeed with the SSL certificate.

To resolve this issue, I moved cacert.pem to the public directory and this is my MYSQL_ATTR_SSL_CA in the ```.env```
```
MYSQL_ATTR_SSL_CA=cacert.pem 
```
This solution solved the problem for me.

In short, Windows users need to download the SSL certificate and place it in the public directory rather than the root directory.

Thank you for your assistance in resolving this matter :)

**Reply by SeakMengs** [reply]
This is just my opinion for improvement I do not have deep knowledge about these stuff but if the SSL certificate path is not correct it should have shown some error about the SSL certificate like the below error message (I intentionally comment out ```MYSQL_ATTR_SSL_CA``` to show the error)
![image](https://github.com/planetscale/discussion/assets/54373229/56ea8728-1ff3-4dd3-a7e3-b4e7d6e8175b)
instead of ```SQLSTATE[HY000] [2002]``` because it leave zero clue on incorrect SSL certificate path.

**Reply by orware** [reply]
Hello @SeakMengs,

Nice work in figuring out this was a certificate issue after all and for sharing the extra details!

I likely would have expected a relative path to work as you originally expected it to, but it is a little odd that the public directory works in this case, but that may be due to it falling back to the web server's public base path for the site in this situation perhaps.

That doesn't seem like the ideal location to place it however. It does look like [absolute paths should generally work](https://dev.to/eriesgo/dotenv-and-relative-paths-fp2) though in the various dotenv libraries that are out there.

I believed relative paths would generally work within an environment file, but I guess it depends on how that value may be used in the code and in this case it is being passed in directly within the `config/database.php` configuration file:

```php
'options' => extension_loaded('pdo_mysql') ? array_filter([
    PDO::MYSQL_ATTR_SSL_CA => env('MYSQL_ATTR_SSL_CA'),
]) : [],
```

I ran a few extra tests and if you would prefer to have it in the root location you can probably try out the following instead within the `config/database.php` file to allow the path to be interpreted relative to the application's root directory using the [base_path()](https://laravel.com/docs/10.x/helpers#method-base-path) helper:

```php
'options' => extension_loaded('pdo_mysql') ? array_filter([
    PDO::MYSQL_ATTR_SSL_CA => base_path() . '/' . env('MYSQL_ATTR_SSL_CA'),
]) : [],
```

When an incorrect certificate path is being provided though, a better error message should probably be displayed but that would be input that would be best shared with the Laravel team in this case perhaps since we don't currently have control over that error message output.

---

#### Tests:

On my end when I tried out the same tests you did, when the provided path wasn't pointing to the file correctly I received the following error, which was a little different than the first error you had experienced and shared above:
```
SQLSTATE[HY000] [2002] (trying to connect via (null))
```

![image](https://github.com/planetscale/discussion/assets/931650/9d26f566-878d-4371-b028-738a9a10bbf1)

Only when I commented out the `MYSQL_ATTR_SSL_CA` variable completely, would I get the clearer error message:

```
SQLSTATE[HY000] [1105] unknown error: Code: UNAVAILABLE server does not allow insecure connections, client must use SSL/TLS
```
